core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java

```diff
@@ -179,6 +180,78 @@ public int getAddressIndex(int position, Page hashChannelsPage)
         return -1;
     }
 
+    @Override
+    public int[] getAddressIndex(int[] positions, Page hashChannelsPage, long[] rawHashes)
+    {
+        return getAddressIndex(positions, hashChannelsPage);
+    }
+
+    @Override
+    public int[] getAddressIndex(int[] positions, Page hashChannelsPage)
+    {
+        checkArgument(hashChannelsPage.getChannelCount() == 1, "Non-signle channel page passed to BigintPagesHash");
+
+        int positionCount = positions.length;
+        long[] incomingValues = new long[positionCount];
+        int[] hashPositions = new int[positionCount];
+
+        for (int i = 0; i < positionCount; i++) {
+            incomingValues[i] = hashChannelsPage.getBlock(0).getLong(positions[i], 0);
+            hashPositions[i] = getHashPosition(incomingValues[i], mask);
+        }
+
+        int[] found = new int[positionCount];
+        int foundCount = 0;
+        int[] result = new int[positionCount];
+        Arrays.fill(result, -1);
+        int[] foundKeys = new int[positionCount];
+
+        // Search for positions in the hash array. This is the most CPU-consuming part as
+        // it relies on random memory accesses
+        for (int i = 0; i < positionCount; i++) {
+            foundKeys[i] = keys[hashPositions[i]];
+        }
+        // Found positions are put into `found` array
+        for (int i = 0; i < positionCount; i++) {
+            if (foundKeys[i] != -1) {
+                found[foundCount++] = i;
+            }
+        }
+
+        // At this step we determine if the found keys were indeed the proper ones or it is a hash collision.
+        // The result array is updated for the found ones, while the collisions land into `remaining` array.
+        int[] remaining = found; // Rename for readability
+        int remainingCount = 0;
+
+        for (int i = 0; i < foundCount; i++) {
+            int index = found[i];
+            if (values[foundKeys[index]] == incomingValues[index]) {
+                result[index] = foundKeys[index];
+            }
+            else {
+                remaining[remainingCount++] = index;
+            }
+        }
+
+        // At this point for any reasoable load factor of a hash array (< .75), there is no more than
+        // 10 - 15% of positions left. We search for them in a sequential order and update the result array.
+        for (int i = 0; i < remainingCount; i++) {
```

Comment:
It will get inlined anyway. As of readability, I prefer those perf-optimised methods to be a bit longer but with some comments. I believe it is more readable that the standard clean code approach, given that the code is not easy to understand and never will be.

Reply:
> It will get inlined anyway

Probably, but profilers can figure out the original method in most cases and attribute the time correctly e.g. in the flame graph.

Reply:
WDYM? If the method is inlined it does not exist in the JFR. Or am I missing something?